Desktop session recording/playback

zmb3 opened the pull request:
Here we introduce a new protobuf type (DesktopRecording) that contains
an encoded TDP message, and update AuditWriter to treat these similarly
to SessionPrint events (which are used for SSH session recordings).

Additionally, a new API server endpoint streams TDP messages from these events
over a websocket interface where they can be rendered by the browser.

Updates #9978
Updates #9979

zmb3 committed:
Record desktop sessions

Here we introduce a new protobuf type (DesktopRecording) that contains
an encoded TDP message, and update AuditWriter to treat these similarly
to SessionPrint events (which are used for SSH session recordings).

We also add desktop session playback endpoint, temporarily located at
/webapi/sites/:site/desktopplaybacktest/:session
which streams TDP messages from a recorded session over
a websocket interface.

zmb3 commented on lib/events/auditwriter.go:
```diff
@@ -139,6 +145,32 @@ func (cfg *AuditWriterConfig) CheckAndSetDefaults() error {
 	if cfg.BackoffDuration == 0 {
 		cfg.BackoffDuration = defaults.NetworkBackoffDuration
 	}
+	if cfg.MakeEvents == nil {
```
Note: we technically don't need this, as it's only used when you use `AuditWriter`'s `io.Writer` interface, and we are calling `EmitAuditEvent()` directly.

However, it felt like a trap to have a general purpose tool called `AuditWriter` that had special case logic for SSH sessions and didn't provide a way to handle desktop or other types of sessions.

I could be convinced to remove this. Longer term, we should probably have a separate type that wraps `AuditWriter` and provides the `io.Writer` interface where necessary.

timothyb89 replied:
I don't especially mind this but it might be nice to keep the default behavior in a named function rather than embedded in a closure here, needing to keep a field `nil` to select a particular implementation seems like an unusual API convention.

codingllama replied:
Seems reasonable to me too, although I'm not intimately familiar with these parts. A named func would be nice (and testable!)

ibeckermayer commented on lib/srv/desktop/tdp/conn.go:
```diff
@@ -29,6 +29,11 @@ import (
 	bufr *bufio.Reader
+
+	// OnSend is an optional callback that is invoked when a TDP message
+	// is sent on the wire. It is passed both the raw bytes and the encoded
```
```suggestion
	// is sent on the wire (when OutputMessage is called). It is passed both the raw bytes and the encoded
```

Commit:
update session end (#9795)

* Updates SessionEnd event with fields needed for frontend

* removes the clock which didn't need to be passed

Commit:
Add `Recorded` field to `WindowsDesktopSessionEnd` (#9839)

* Adds the SessionRecording field to WindowsDesktopSessionEnd event to mimic SessionEnd events (useful for easy integration with frontend).

* 14 should have been 12

* removing test logic

* switches SessionRecording to simple boolean Recorded

Commit:
session recording websocket (#9908)

* Adds the SessionRecording field to WindowsDesktopSessionEnd event to mimic SessionEnd events (useful for easy integration with frontend).

* 14 should have been 12

* removing test logic

* switches SessionRecording to simple boolean Recorded

* Updates the websocket address

* updates desktopPlaybackHandle to restart playback once it reaches the end

* adds playback state and synchronization logic for ensuring that goroutines aren't leaked

* adds toggle functionality for play/pause

* fix for the fact that urls are case insensitive

* moves desktop_playback to its own file, fixes mistaken comment about how websocket.JSON.Receive works, fixes error messaging, wraps playbackState.Close in a sync.Once

* Adds a cond variable for the two goroutines, but doesn't solve the spinning loop problem for the hanging logic

* Adds a cancel-able context which is cancelled in ps.Close() in order to avoid a spinning loop in the websocket.Handler

* Moves the majority of playback logic into the playbackState, which is now renamed to the more accurate playbackPlayer.

* changes pp.hangWhilePaused to pp.waitWhilePaused

* Moves the context out of NewPlaybackPlayer and the playbackPlayer
struct, wraps playback goroutines in playbackPlayer.Play(ctx) in
order to comply with context semantics.

* removing unnecessary warnings

* touchups

Commit:
record screen size (#9992)

* Adds the SessionRecording field to WindowsDesktopSessionEnd event to mimic SessionEnd events (useful for easy integration with frontend).

* 14 should have been 12

* removing test logic

* switches SessionRecording to simple boolean Recorded

* Updates the websocket address

* updates desktopPlaybackHandle to restart playback once it reaches the end

* adds playback state and synchronization logic for ensuring that goroutines aren't leaked

* adds toggle functionality for play/pause

* fix for the fact that urls are case insensitive

* moves desktop_playback to its own file, fixes mistaken comment about how websocket.JSON.Receive works, fixes error messaging, wraps playbackState.Close in a sync.Once

* Adds a cond variable for the two goroutines, but doesn't solve the spinning loop problem for the hanging logic

* Adds a cancel-able context which is cancelled in ps.Close() in order to avoid a spinning loop in the websocket.Handler

* Moves the majority of playback logic into the playbackState, which is now renamed to the more accurate playbackPlayer.

* changes pp.hangWhilePaused to pp.waitWhilePaused

* Moves the context out of NewPlaybackPlayer and the playbackPlayer
struct, wraps playback goroutines in playbackPlayer.Play(ctx) in
order to comply with context semantics.

* removing unnecessary warnings

* Adds an OnRecv that's similar to OnSend, for emitting audit events for particular incoming tdp messages

Commit:
Send full `DesktopRecording` event as json over playback websocket. (#10052)

* playback websocket now sends a json representation of the DesktopRecording event rather than just the raw tdp message, in order for us to have timing data on the frontend

* updating json.Marshal to utils.FastMarshal

* Removing unnecessary comment

Commit:
playback end event (#10088)

* Adds an end event so that the playback player knows to set the progress bar to its end state

* making the end message a json

* if the marshal fails we don't want to send a message over websocket

* Use a static string

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

zmb3 committed:
Add participants to session end event

Desktop sessions are not joinable, so the participants list always
has a single member - the user who started the session.

This will ensure that our example role for RBAC for sessions
(which depends on the participants field) will work for desktop
sessions.

zmb3 committed:
Minor cleanup

zmb3 committed:
Only record sessions when enabled

In order for desktop sessions to be recorded, session recording
must be enabled in the cluster's session recording config and
at least one of the user's roles must enable it.

zmb3 committed:
Cleanup

codingllama commented on lib/srv/desktop/windows_server.go:
```diff
@@ -333,6 +340,7 @@ func NewWindowsService(cfg WindowsServiceConfig) (*WindowsService, error) {
 		Clock: s.cfg.Clock,
 	})
 	if err != nil {
+		s.Close()
```
This might be a good place to apply that "!ok/close" idiom:

```go
// Declare closeable resources
ok := false
defer func() {
  if !ok {
    // Close resources
  }
}()

// Method continues ...

ok = true
return
```

zmb3 replied:
I like it!

codingllama commented on lib/srv/desktop/windows_server.go:
```diff
@@ -644,6 +709,33 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
+	}
+
+	defer func() {
+		go func() {
```
Maybe leave a comment here? A `go` inside a `defer` is not that common.

codingllama commented on lib/srv/desktop/windows_server.go:
```diff
@@ -660,7 +752,45 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
+	delay := timer()
 	tdpConn := tdp.NewConn(proxyConn)
+	sessionStartTime := s.cfg.Clock.Now().UTC().Round(time.Millisecond)
+	tdpConn.OnSend = func(m tdp.Message, b []byte) {
```
Maybe give the event handlers names so there is less noise inside connectRDP? That way you could also test them separately.

codingllama commented on lib/srv/desktop/windows_server.go:
```diff
@@ -660,7 +752,45 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
+	}
+	tdpConn.OnRecv = func(m tdp.Message) {
+		switch m.(type) {
+		// TODO: tdp.ClipboardData
```
```suggestion
		// TODO(zmb3): Record audit receive events for tdp.ClipboardData messages
```

codingllama commented on lib/srv/desktop/windows_server.go:
```diff
@@ -660,7 +752,45 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
+		case tdp.ClientScreenSpec, tdp.MouseButton, tdp.MouseMove:
+			b, err := m.Encode()
+			if err != nil {
+				s.cfg.Log.WithError(err).Warning("could not emit desktop recording event")
```
Should we record the fact that the encode operation failed? It is worth having that and/or the error in the event?

zmb3 replied:
As of now, this can basically never happen, as encode just writes to a `*bytes.Buffer`, which is guaranteed to always return a `nil` error.

codingllama commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,246 @@
+/*
+Copyright 2021 Gravitational, Inc.
```
```suggestion
Copyright 2022 Gravitational, Inc.
```

I would only bother with new files, though.

codingllama commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,246 @@
+	"golang.org/x/net/websocket"
+)
+
+const actionPlayPause = "play/pause"
+
+const (
+	playStatePlaying  = "playing"
+	playStatePaused   = "paused"
+	playStateFinished = "finished"
```
Do we want to strongly type these? (Instead of string.)

codingllama commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,246 @@
+// to the server over the websocket connection in order to modify
+// the playback state.
+type playbackAction struct {
+	// Action is one of actionPlayPause | TODO: actionMove
```
This comment looks a bit off. Move the TODO to another line and the part about actionPlayPause to the const itself?

codingllama commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,246 @@
+/*
```
Do we have a better package for this file than lib/web? I'm trying to avoid adding to the more "general" packages.

codingllama replied:
I like that too, sounds good.

codingllama commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,246 @@
+}
+
+// playbackPlayer is a thread-safe struct for managing
+// the global state of the playback websocket connection.
```
What is the "playback websocket connection"? Could you give it more context?

codingllama commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,246 @@
+
+// playbackPlayer is a thread-safe struct for managing
+// the global state of the playback websocket connection.
+type playbackPlayer struct {
```
Does this plug into anything? Should it be in another PR?

Could we have tests too?  ;)

codingllama commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,246 @@
+// Play kicks off goroutines for receiving actions
+// and playing back the session over the websocket,
+// and then hangs until the websocket is closed.
+func (pp *playbackPlayer) Play(ctx context.Context) {
```
Move this higher to the top? The rest made a lot more sense once I read Play.

codingllama commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,246 @@
+func (pp *playbackPlayer) Play(ctx context.Context) {
+	defer pp.log.Debug("playbackPlayer.Play returned")
+	ppCtx, cancel := context.WithCancel(ctx)
+	defer pp.close(cancel)
```
I found passing an external cancel like this a bit strange, specially when reading the methods that take CancelFunc as an argument.

If we are not allowing players to be reused (which I think is the case), then maybe consider having an internal cancel that the funcs can access?

zmb3 replied:
This came about because I'm not a fan of holding on to `context.Context` (or its associated cancel funcs) as struct members and asked Isaiah to see if we could pass them as arguments instead.

In this case, I guess we could capture the CancelFunc as a member of the player struct, but my newfound Rust experience is telling me this is a bad idea. The context is only valid for the duration of `Play`, so we shouldn't stash the associated cancel somewhere that could outlive `Play`.

Have any ideas?

codingllama replied:
If we don't allow more than a single invocation of Play, which seems to be the intent, then my initial idea would be to nest playCtx/playCancel fields inside playbackPlayer, initialized on new.

In general I don't like concurrency primitives passed around - funcs with channels, locks or cancels as params. I'd either handle concurrency where the goroutines are launched (ie, sync primitives live in a single method) or nest attributes. That said, what you have definitely works and it's all private, so feel free to push back if you don't think this is worth pursuing.

zmb3 replied:
I don't love either approach, to be honest. Right now, we don't allow more than one invocation of `Play()`. But that might change. 

I'd prefer that we continue to pass the context to play, as that is the long running operation that must support cancelation, and I can't really think of a better way to maintain that invariant without passing around a shared cancelation signal.

codingllama reviewed (commented):
Sorry about the delay, reviewed.

zmb3 commented on api/types/events/events.proto:
```diff
@@ -230,6 +230,20 @@ message SessionPrint {
+    bytes Message = 2 [ (gogoproto.nullable) = true, (gogoproto.jsontag) = "message" ];
+
+    // DelayMilliseconds is the delay in milliseconds from the start of the session
+    int64 DelayMilliseconds = 3 [ (gogoproto.jsontag) = "ms" ];
```
I agree. In this case we were trying to keep compatibility with `message SessionPrint` (which is the equivalent protobuf for SSH sessions).

Given that context, are you okay with keeping this or still want to see it changed?

In all current use cases, the front end is either expecting SSH or desktop, so they don't necessarily need to match, but was being cautious as we didn't know how the front end would play out when we first added recording.

codingllama replied:
Right, I thought it might be something like this. Up to you, but maybe leave a comment so context is not lost?

zmb3 replied:
fair, will do

zmb3 commented on api/types/events/events.proto:
```diff
@@ -1676,6 +1690,24 @@ message WindowsDesktopSessionEnd {
+    google.protobuf.Timestamp StartTime = 9 [
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "session_start,omitempty"
```
Same here, following the unfortunate precedent set by `message SessionEnd`.

zmb3 commented on lib/events/api.go:
```diff
@@ -453,9 +453,12 @@ const (
+	// WindowsDesktopSessionEndEvent is emitted when a user disconnects
 	// from a desktop.
 	WindowsDesktopSessionEndEvent = "windows.desktop.session.end"
+
```
Eh.`DesktopRecordingEvent` is OS-agnostic, and would work if we end up supporting desktops on macOS or Linux. This is why it was not grouped with the Windows-specific events.

codingllama replied:
Makes sense, my bad.

zmb3 committed:
Start to address review comments

zmb3 committed:
Move TDP event handlers out of connectRDP

codingllama commented on lib/web/desktop/playback.go:
```diff
@@ -0,0 +1,240 @@
+// Player manages the playback of a recorded desktop session.
+// It streams events from the audit log to the browser over
+// a websocket connection.
+type Player struct {
```
Oh yes. Separate package, public methods make a coherent interface, looks great. :D

codingllama reviewed (commented):
Give me a nudge when you are done with the comments.

zmb3 committed:
Address more review comments and add some tests

zmb3 committed:
Add playback streaming test

zmb3 committed:
Consistent comments

zmb3 reviewed (commented):
@codingllama mind taking one more pass?

I added a simple streaming test. We'll be moving from `x/net/websocket` to `gorilla/websocket` after this PR, at which time I'll be able to test more sophisticated error and cancelation scenarios.

zmb3 commented on lib/srv/desktop/windows_server_test.go:
```diff
@@ -192,3 +195,83 @@ func TestGenerateCredentials(t *testing.T) {
+		dr, ok := e.(*events.DesktopRecording)
+		require.True(t, ok)
+		require.Equal(t, encoded, dr.Message)
+	case <-time.After(1 * time.Second):
```
1 second is an eternity for a channel send. I was aiming for a quick failure here rather than a failure that causes CI to to hang until the max timeout. 🤷‍♂️

codingllama replied:
Fair.

zmb3 committed:
Fix tests

zmb3 committed:
Don't log PNG frames that exceed the size of a protobuf

Since the PNG frame message in our desktop protocol is unbounded,
it is theoretically possible for a message to exceed the size limit
of a single protobuf.

In practice, this is unlikely to occur with any legitimate RDP traffic,
as the bitmaps are at most 64x64 pixels and compressed in PNG form.
Rather than complicating the protocol to allow for PNGs to be split
across events, we simply refuse to log anything this big.

zmb3 committed:
Mark RFD 48 implemented
